drivers/DeepSmart/deepsmart/src/ssdp.lua

```diff
@@ -0,0 +1,89 @@
+local socket = require('socket')
+local log = require('log')
+local config = require('config')
+
+
+
+
+local SSDP = {}
+
+local DEEPSMART_SSDP_SEARCH_TERM = "DEEPSMART-ARM"
+-----------------------
+-- SSDP Response parser
+local function parse_ssdp(data)
+  local res = {}
+  res.status = data:sub(0, data:find('\r\n'))
+  for k, v in data:gmatch('([%w-]+): ([%a+-: /=]+)') do
+    local key = k:lower()
+    log.info('parse key '..key..' val '..v)
+    res[key] = v
+  end
+  return res
+end
+
+
+-- This function enables a UDP
+-- Socket and broadcast a single
+-- M-SEARCH request, i.e., it
+-- must be looped appart.
+function SSDP.search(callback)
+  local bridges = {}
+  -- UDP socket initialization
+  local upnp = socket.udp()
+  local _,err = upnp:setsockname('0.0.0.0', 0)
+  if err then
+    log.error(string.format("udp socket failure setsockname: %s", err))
+    return false,bridges,err
+  end
+  -- use broadcast to find wisers as wifi router's group cast is forbidden by default(need login to router to start group cast)
+  upnp:setoption('broadcast', true)
+  local timeout = socket.gettime() + config.MC_TIMEOUT + 1
+  local mx = 2
+  -- broadcasting request
+  log.info('===== SCANNING NETWORK...')
+  local multicast_msg = table.concat({
+    "M-SEARCH * HTTP/1.1",
+    "HOST: 255.255.255.255:1900",

```

Comment:
use broadcast to search bridge while wifi routers will disable group cast by default

Reply:
Can you explain why this is needed? Was it just one network that had multicast disabled? We have used multicast rather than broadcast for all other SSDP use cases, and have not had a problem or user complaint; also the SSDP rfc specifies the multicast address 239.255.255.250 should be used for SSDP. 

Making this change adds more concerns that the driver's discovery and periodic IP address checks will have adverse network affects in large networks.

Reply:
The on-site router model does not support multicast

Reply:
Did this actually fix the issue? The `search-parameters.yml` file specifies the ssdp search term, and the hub uses the multicast address, so if devices are not responding to that request, the driver will never start.

Reply:
Only when the gateway responds to the ssdp query request, the device does not need to respond. ssdp will search for all devices on the gateway after finding it

Reply:
If the gateway responds to the multicast request (ssdp query), why would the broadcast need to be used. Also, FWIW broadcast is not a socket option that we allow for the sockets in lua drivers, which I believe is needed to the broadcast addr.

Reply:
Gateway can respond to both broadcast and multicast ssdp requests
Many WiFi routers have disabled multicast, so they can only send ssdp request data through broadcasting

Reply:
> Many WiFi routers have disabled multicast, so they can only send ssdp request data through broadcasting

The reason I dont understand this change, is that if a gateway has disabled multicast, the driver will not start. The hub uses the ssdp search term in `search-paremeters.yml` with the multicast address to determine if the driver will be started. So if that is disabled on the gateway, the hub will never think there is a device on the network, and this driver will never be run.

Reply:
1. The router does not allow multicast, and the hub cannot find the gateway ；
2. The driver will run, but the driver cannot search for the gateway ；

Reply:
After our testing cases and testing process, the problem you mentioned does not exist. Our testing results are all okay

Reply:
> After our testing cases and testing process, the problem you mentioned does not exist. Our testing results are all okay

Hi @cjswedes 

DeepSmart claims that anything goes well. 
If there are no other unresolved issues, can we approve this PR first and let other departments test it?

We have to a get an approval of this PR, then ask other departments to do certification test.